src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+            nextNode.previous = currentNode;
+            head = currentNode;
+            size++;
+        } else {
+            currentNode = new Node<>(nextNode.previous, nextNode, value);
+            nextNode.previous.next = currentNode;
+            nextNode.previous = currentNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        ioobException(index);
+        return findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        ioobException(index);
+        Node<T> newNode = findByIndex(index);
+        T oldVal = newNode.value;
+        newNode.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        ioobException(index);
+        Node<T> removable = findByIndex(index);
+        unlink(removable);
+        size--;
+        return removable.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removable = head;
+        for (int i = 0; i < size; i++) {
+            if (object == removable.value
+                    || (object != null && object.equals(removable.value))) {
+                remove(i);
+                return true;
+            } else {
+                removable = removable.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void unlink(Node<T> node) {
+        Node<T> previousNode = node.previous;
+        Node<T> nextNode = node.next;
+        if (previousNode == null && nextNode != null) {
+            nextNode.previous = null;
+            head = nextNode;
+        }
+        if (nextNode == null && previousNode != null) {
+            previousNode.next = null;
+            tail = previousNode;
+        }
+        if (previousNode != null && nextNode != null) {
+            previousNode.next = nextNode;
+            nextNode.previous = previousNode;
+        }
+    }
+
+    private Node<T> findByIndex(int index) {
+        ioobException(index);
+        if (index == 0) {
+            return head;
+        }
+        if (index == size - 1) {
+            return tail;
+        }
+        Node<T> indexedNode;
+        if (index < size / 2) {
+            indexedNode = head;
+            for (int i = 0; i < index; i++) {
+                indexedNode = indexedNode.next;
+            }
+        } else {
+            indexedNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                indexedNode = indexedNode.previous;
+            }
+        }
+        return indexedNode;
+    }
+
+    private void ioobException(int index) {
```

Comment:
Exception is a class, but we have the method. I think you should rename this method like checkIndex() and refactor code where this method called
```suggestion
    private void checkIndex(int index) {
```